[#718] Generate Public Certificate

sam-bertin opened the pull request:
- Added a section to the CSR model that allows for public certificate upload -Uploaded certificate is checked against the private key to make sure it's valid
- valid certificate gets added as an instance with it's private key
- admin_client test to geneneate the certificate from the CSR admin model change view.
- Test edusing the admin client

sam-bertin committed:
[#718] 🧬 Generate Public Certificate

Handle public certificate upload (certificate creation :

- Added a section to the CSR model that allows for public certificate upload
-Uploaded certificate is checked against the private key to make sure it's valid
- valid certificate gets added as an instance with it's private key
- admin_client test to generate the certificate from the CSR admin model change view.
- Re uploading a public certificate does not create a new instance 
- Updated docs with the new features
- replace exclude from Form with fields=_all__

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -79,8 +79,30 @@ class SigningRequestAdmin(admin.ModelAdmin):
+        ),
     )
-    readonly_fields = ("csr",)
+    list_display = (
+        "common_name",
+        "organization_name",
+        "country_name",
+        "state_or_province_name",
+        "locality_name",
+        "email_address",
+    )
+    list_filter = ("organization_name", "state_or_province_name", "locality_name")
+    search_fields = ("common_name", "organization_name", "locality_name")
```
this is already defined earlier and now duplicated - probably something went wrong while resolving a merge conflict?

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -15,6 +19,13 @@
 class SigningRequestAdminForm(forms.ModelForm):
+    certificate = forms.FileField(
+        required=False,
+        help_text="Upload the public certificate file here. "
+        "This will be used to verify the signature against the CSR "
+        "and create the certificate instance.",
+        label="Upload Certificate",
```
1. user-facing text must be marked translatable
2. I think we can/should simplify the help text: "Upload the certificate file received from your certificate authority matching the CSR you sent them."

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+        if self.instance.public_certificate:
+            self.add_error(
+                "certificate",
+                "A certificate already exists for this CSR. "
+                "Delete the certificate first.",
```
user-facing text must be marked translatable

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
 
+    def clean(self) -> dict[str, Any]:
+        # Errors if the CSR already has a certificate
+        if self.instance.public_certificate:
+            self.add_error(
+                "certificate",
+                "A certificate already exists for this CSR. "
+                "Delete the certificate first.",
+            )
```
I think this is a check you can put in `def clean_certificate`, it doesn't depend on other form fields

sam-bertin replied:
this I did not manage to make it work. I wrote smth like :

```
    def clean_certificate(self) -> File | None:
        # Errors if the CSR already has a certificate
        if self.instance.public_certificate:
            self.add_error(
                "certificate",
                _(
                    "A certificate already exists for this CSR. "
                    "Delete the certificate first.",
                ),
            )
        return self.cleaned_data.get("certificate")
```

But then I've got empty files when uploading the cert.

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -35,8 +46,12 @@ def __init__(self, *args, **kwargs):
+    def save(self, commit: bool = True) -> Any:
+        instance = super().save(commit)
+        # Handle the certificate file upload
+        validated_certificate = self.cleaned_data.get("certificate")
+        if validated_certificate:
+            instance.create_certificate(validated_certificate)
         if self.cleaned_data.get("should_renew_csr", False):
```
Reading this code, it looks like I can now upload a certificate AND generate a new private key/CSR at the same time. That should be prevented:

1. you either generate a new private key/CSR and then no certificate file may be uploaded
2. or you upload a certificate file, and then `should_renew_csr` must be `False`.

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+                "Delete the certificate first.",
+            )
+        # Handle the certificate file upload
+        if "certificate" in self.cleaned_data and self.cleaned_data["certificate"]:
+            uploaded_file = self.cleaned_data["certificate"]
```
```suggestion
        if (uploaded_file := self.cleaned_data.get("certificate")):
```

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+                certificate_data = uploaded_file.read()
+
+                # Load the certificate and throw an error if the certificate is invalid
+                certificate = x509.load_pem_x509_certificate(certificate_data)
```
What happens when the certificate is not a PEM file? There is no guarantee that the CA will provide these files as PEM data, often they are some binary format (PKCS?)

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+
+                # Load the certificate and throw an error if the certificate is invalid
+                certificate = x509.load_pem_x509_certificate(certificate_data)
+                private_key = load_pem_x509_private_key(
+                    self.instance.private_key.encode()
+                )
```
I think everywhere else we explicitly encode as `ascii` - please don't rely on implicit encoding parameters when converting to/from binary data since it can lead to bugs that are hard to track down.

I believe that the Python 2 -> Python 3 transition changed the default value of the encoding, so making it explicit is much more future proof.

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+                if not match:
+                    self.add_error(
+                        "certificate",
+                        "Certificate does not match the signature from the actual CSR.",
```
user-facing text must be marked translatable

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+                        "Certificate does not match the signature from the actual CSR.",
+                    )
+
+            except ValueError:
+                self.add_error(
+                    "certificate", "Invalid certificate. Check the file format."
```
user-facing text must be marked translatable

sergei-maertens replied:
there are many lines in the `try` block that can raise a `ValueError`, e.g. loading a private key with password

Reading the file contents is also not something that should be inside this `try` block, we shouldn't expect that to error otherwise we need to file a bug report with Django.

Try to wrap only the code you expect to raise an exception, so rather do this instead:

```py
uploaded_file.seek(0)
certificate_data = uploaded_file.read()
try:
    certificate = x509.load_pem_x509_certificate(certificate_data)
except ValueError:
    ...

# not wrapped, we control the private key and failing to load it means a bug in
# our code that needs fixing, the end-user cannot solve any validation errors we
# show
private_key = load_pem_x509_private_key(self.instance.private_key.encode())

... # validate public keys against each other etc. etc.
```

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
+                self.add_error(
+                    "certificate", "Invalid certificate. Check the file format."
+                )
+            except Exception:
+                self.add_error("certificate", "An unexpected error occurred.")
```
```suggestion
```
it's a bad idea to display programming errors to the user as if they're form validation errors.

If an unexpected error happens, then we need to see that it and what happened in Sentry, and suppressing exceptions is usually a bad idea for that reason.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -1,5 +1,7 @@
 from datetime import datetime
+from io import StringIO
 
+from django.core.files.base import File
```
```suggestion
from django.core.files import File
```
I *think* the parent module already exports it

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -60,6 +70,21 @@ class Meta:
             )
         ]
 
+    def create_certificate(self, certificate):
```
```suggestion
    def create_certificate(self, certificate: File):
```

sergei-maertens replied:
```suggestion
    @transaction.atomic
    def create_certificate(self, certificate):
```
since you're doing multiple writes to the DB in a single method, best to wrap it in an atomic block so that all or nothing succeeds if there are errors

sergei-maertens replied:
following the django code style I linked earlier, custom methods go _below_ the `__init__` and `save` (and other built in methods)

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -241,3 +246,193 @@ def test_admin_save_existing_csr_should_not_renew(admin_client):
+        email_address="email@valid.com",
+    )
+
+    cert_pem = ContentFile(b"invalid bytes", "cert.pem")
```
could also be a `SimpleUploadedFile`, it's a bit odd that different file classes are used now in tests

```suggestion
    cert_pem = SimpleUploadedFile("cert.pem", b"invalid bytes")
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -79,8 +79,30 @@ class SigningRequestAdmin(admin.ModelAdmin):
             },
         ),
+        (
+            _("Upload and verify certificate"),
```
```suggestion
            _("Upload certificate"),
```

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -79,8 +79,30 @@ class SigningRequestAdmin(admin.ModelAdmin):
+            _("Upload and verify certificate"),
+            {
+                "fields": ("certificate", "public_certificate"),
+                "description": _(
+                    "Upload the public certificate file here. "
+                    "This will be used to verify the signature against the CSR "
+                    "and create the certificate instance. The CSR needs to be saved"
+                    " first (created) before uploading the certificate."
+                ),
```
I think the help text on the fields is sufficient, let's drop this.

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -241,3 +246,193 @@ def test_admin_save_existing_csr_should_not_renew(admin_client):
+    )
+
+    private_key = load_pem_x509_private_key(csr.private_key.encode())
+    # Here I did not use the fixture mkcert to generate the certificate
+    # This is because the fixture mkcert did not generate a valid certificate
+    # The public keys of the private key and the certificate did not match
```
the `mkcert` allows us to pass in the subject and issuer key - what exactly is not working? And I'd rather add more options to `mkcert` then instead, now this seems to just be duplicating a bunch of code.

IMO we should be able to at least re-use the `root_key` and `root_cert` fixtures for the `issuer` parameters in `mkcert`.

Maybe an optional `csr` parameter needs to be added so you can properly copy over the subject details from the CSR.

sam-bertin replied:
I don't really remember why I couldn't use it before but I could make it fit pretty easily this time.

sergei-maertens commented on tests/test_csr.py:
```diff
@@ -241,3 +246,193 @@ def test_admin_save_existing_csr_should_not_renew(admin_client):
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.datetime.now())
+        .not_valid_after(datetime.datetime.now() + datetime.timedelta(days=365))
+        .add_extension(
+            x509.SubjectAlternativeName([x509.DNSName("localhost")]),
+            critical=False,
+        )
```
this doesn't really make sense for TLS clients - they don't have a (public) DNS name

sam-bertin committed:
[#718] Process review

- 🚸 Translatable user facing text
- 🔥 Remove duplicated code
- ♻️ Small refactors

alextreme commented:
@CharString please process Sergei's feedback together with Samuel

sam-bertin committed:
[#718] Process Review

🌐 User translatable errors
Forgot to translate some of the user facing strings.

🎨 Explicit ASCII encoding
All encoding of cryptography files can be done with ascii since PEM format is 64-bits.

🎨 Django Coding Styles 
Change the order of methods from models to django coding styles.


🔥 Removed exception from  the form validation errors

🏷️ File typing on create_certificate

🦺 Uploaded certificate validator
Validate the certificate before the clean method. This way there is no need to use the try/except block as the load methods won't throw errors at this point.

💬 Updated test error message.
Since the validation of the cert changed the error message isn't the same.

🏗️ Moved the handling of the certificate in its own formfield cleaning method (clean_certificate)

✅ Updated tests
